File: Data Structures & Algorithms/two-integer-sum/submission-14.cs
public class Solution {
    // I will iterate throught the nums array, getting the difference between the curr num and the target
    // and then storing it in a HashMap where the key is the diff and value is the index
    // As I iterate through the nums I will check if the hashMap contains the diff, if so return the 
    // the index of the current num and the value of the hashMap element. 
    // HashMap is used for O(1) lookups

    // Time Complexity is O(n) where n is the number of elements in the array
    // Space Complexity is O(n) where in worst case we store all the elements in nums in the HashMap

    public int[] TwoSum(int[] nums, int target) {
        var hashMap = new Dictionary<int, int>();

        for (var i = 0; i < nums.Length; i++) {
            var diff = target - nums[i]; // 7 - 3 = 4
            if(hashMap.ContainsKey(diff)) return new int[2] { hashMap[diff], i};
            hashMap.Add(nums[i], i);
        }

        return new int[2];
    }
}

File: Data Structures & Algorithms/two-integer-sum-ii/submission-14.ts
class Solution {
    /**
     * @param {number[]} numbers
     * @param {number} target
     * @return {number[]}
     */
    twoSum(numbers: number[], target: number): number[] {
        let l = 0
        let r = numbers.length - 1

        while (l <= r) {
            const sum = numbers[r] + numbers[l]

            if (sum === target) return [l + 1, r + 1]
            else if (sum > target) r--
            else l++
        }

        return []
    }
}
